openfl/component/director/director.py

```diff
@@ -6,50 +6,279 @@
+from typing import List
+from typing import Union
 
+from openfl.component import Aggregator
 from openfl.federated import Plan
 from openfl.protocols import director_pb2
+from openfl.transport import AggregatorGRPCServer
 from openfl.utilities.workspace import ExperimentWorkspace
 
 logger = logging.getLogger(__name__)
 
 ENVOY_HEALTH_CHECK_PERIOD = 60  # in seconds
 
 
+class Status:
+    """Experiment's statuses."""
+
+    PENDING = 'PENDING'
+    FINISHED = 'FINISHED'
+    IN_PROGRESS = 'IN_PROGRESS'
+    FAILED = 'FAILED'
+
+
+class Experiment:
+    """Experiment class."""
+
+    def __init__(
+            self, *,
+            name: str,
+            data_path: Union[Path, str],
+            collaborators: List[str],
+            sender: str,
+            init_tensor_dict: dict,
+            plan_path: Union[Path, str] = 'plan/plan.yaml',
+            aggregator_grpc_server: AggregatorGRPCServer = None,
+            users: Iterable[str] = None,
+    ) -> None:
+        """Initialize an experiment object."""
+        self.name = name
+        if isinstance(data_path, str):
+            data_path = Path(data_path)
+        self.data_path = data_path
+        self.collaborators = collaborators
+        self.sender = sender
+        self.init_tensor_dict = init_tensor_dict
+        if isinstance(plan_path, str):
+            plan_path = Path(plan_path)
+        self.plan_path = plan_path
+        self.__aggregator_grpc_server = aggregator_grpc_server
+        self.users = set() if users is None else set(users)
+        self.status = Status.PENDING
+
+    @property
+    def aggregator(self) -> Union[Aggregator, None]:
+        """Get aggregator."""
+        if self.__aggregator_grpc_server:
+            return self.__aggregator_grpc_server.aggregator
+
+    async def start(
+            self, *,
+            tls: bool = False,
+            root_certificate: Union[Path, str] = None,
+            certificate: Union[Path, str] = None,
+            private_key: Union[Path, str] = None,
+    ) -> None:
+        """Start the experiment."""
+        self.status = Status.IN_PROGRESS
+        try:
+            logger.info(f'New experiment {self.name} for '
+                        f'collaborators {self.collaborators}')
+
+            await self._run_aggregator(
+                tls=tls,
+                root_certificate=root_certificate,
+                certificate=certificate,
+                private_key=private_key,
+            )
+            self.status = Status.FINISHED
+        except Exception:
+            self.status = Status.FAILED
+
+    async def _run_aggregator(
+            self, *,
+            tls: bool = False,
+            root_certificate: Union[Path, str] = None,
+            certificate: Union[Path, str] = None,
+            private_key: Union[Path, str] = None,
+    ) -> None:
+        with ExperimentWorkspace(self.name, self.data_path):
+            self.__aggregator_grpc_server = self._create_aggregator_grpc_server(
```

Comment:
No reason to return aggregator_grpc_server, it's internal method and it can define self.__aggregator_grpc_server by itself. Or we should pass it into  _run_aggregator_server,  because _run_aggregator_server will be failed without running self._create_aggregator_grpc_server.

Reply:
Done